mentoring/mentoring.py

```diff
@@ -104,61 +260,297 @@ def next_step_url(self):
+
+        self.runtime.publish(self, 'xblock.mentoring.assessment.submitted', event_data)
+
+        return {
+            'completed': completed,
+            'attempted': self.attempted,
+            'max_attempts': self.max_attempts,
+            'num_attempts': self.num_attempts,
+            'step': self.step,
+            'score': score.percentage,
+            'correct_answer': score.correct,
+            'incorrect_answer': score.incorrect,
+            'partially_correct_answer': score.partially_correct,
+        }
+
+    @XBlock.json_handler
+    def try_again(self, data, suffix=''):
+
+        if self.max_attempts_reached:
+            return {
+                'result': 'error',
+                'message': 'max attempts reached'
+            }
+
+        # reset
+        self.step = 0
+        self.completed = False
+
+        while self.student_results:
+            self.student_results.pop()
+
+        return {
+            'result': 'success'
+        }
+
+    @property
+    def max_attempts_reached(self):
+        return self.max_attempts > 0 and self.num_attempts >= self.max_attempts
 
     def get_message_html(self, message_type):
-        fragment = self.get_message_fragment(message_type)
-        if fragment:
-            return fragment.body_html()
-        else:
-            return ''
+        html = u""
+        for child_id in self.children:
+            child = self.runtime.get_block(child_id)
+            if isinstance(child, MentoringMessageBlock) and child.type == message_type:
+                html += child.render('mentoring_view', {}).content  # TODO: frament_text_rewriting ?
+        return html
+
+    def clean_studio_edits(self, data):
+        """
+        Given POST data dictionary 'data', clean the data before validating it.
+        e.g. fix capitalization, remove trailing spaces, etc.
+        """
+        if data.get('mode') == 'assessment' and 'max_attempts' not in data:
+            # assessment has a default of 2 max_attempts
+            data['max_attempts'] = 2
+
+    def validate(self):
+        """
+        Validates the state of this XBlock except for individual field values.
+        """
+        validation = super(MentoringBlock, self).validate()
+        a_child_has_issues = False
+        message_types_present = set()
+        for child_id in self.children:
+            child = self.runtime.get_block(child_id)
+            # Check if the child has any errors:
+            if not child.validate().empty:
+                a_child_has_issues = True
+            # Ensure there is only one "message" block of each type:
+            if isinstance(child, MentoringMessageBlock):
+                msg_type = child.type
+                if msg_type in message_types_present:
+                    validation.add(ValidationMessage(
+                        ValidationMessage.ERROR,
+                        self._(u"There should only be one '{}' message component.".format(msg_type))
```

Comment:
This validation message remains static even when this issue is resolved. Validation should refresh when components are added or removed.